src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+    tmp_ptr = tmp_ptr->next;
+  }
 
+  // Check whether the server is overloaded
+  if ((!queue && client_count >= 1) ||
+      (queue && !multiple && client_count >= max_clients) ||
+      (queue && multiple && client_count >= 4 * max_clients)) {
+    // The server is overloaded. Reject the client.
+    log_println(0,
+                "Too many clients/mclients (%d) waiting to be served, "
+                "Please try again later.",
+                new_child->pid);
+    child_signal = SRV_QUEUE_SERVER_BUSY;
+    write(new_child->pipe, &child_signal, sizeof(ParentChildSignal));
+    // Whether the write succeeds or not, we are done with this child. It
+    // will either kill itself in response to this message, or it will kill
+    // itself with its own watchdog timer. Either way, not our problem
+    // anymore. Send the message, close the pipe, and move on.
+    close(new_child->pipe);
+    log_println(6, "Too many clients, freeing child=0x%x", new_child);
+    free(new_child);
+    new_child = NULL;
+  } else {
+    // The server is not overloaded, so queue up the new client
+    if (*head == NULL) {
+      *head = *tail = new_child;
     } else {
-      // Nothing is in the queue, so wait forever until a new connection
-      // request arrives
-      log_println(3, "Timer not running, waiting for new connection");
-      mclients = 0;
-      do {
-        retcode = select(listenfd + 1, &rfd, NULL, NULL, NULL);
-        // retry if a signal caused the select() to exit
-      } while ((retcode == -1) && (errno == EINTR));
+      (*tail)->next = new_child;
+      *tail = new_child;
     }
+  }
+}
 
-    if (retcode < 0) {
-      // an interrupt or signal caused the select() to exit, go back and
-      // start over
-      log_println(5, "Select exited with return code = %d", retcode);
-      continue;
-    }
+/**
+ * Returns true or false, depending on whether the client's process is dead.
+ */
+int is_live_client(struct ndtchild *client) {
+  siginfo_t client_status;
+  int rv;
+  client_status.si_pid = 0;
+  rv = waitid(P_PID, client->pid, &client_status, WEXITED | WSTOPPED | WNOHANG);
+  return (rv == 0 && client_status.si_pid == 0);
+}
 
-    if (retcode == 0) {  // select exited due to timer expiration
-      log_println(3, "Timer expired while waiting for a new connection");
-      /* if ((waiting > 0) && (testing == 0)) */
-      if (multiple == 0) {
-        if ((waiting > 0) && (testing == 0)) goto ChldRdy;
-        continue;
-      } else {
-        if (waiting > 0) goto ChldRdy;
-        continue;
-      }
+/** Goes from head to tail, reaping dead clients from the queue of clients and
+ * then updating all clients regarding their (possibly new) queue position. */
+void queue_maintenance(struct ndtchild **head, struct ndtchild **tail) {
+  ParentChildSignal child_message;
+  int queue_position;
+  int client_count = 0;
+  int running_count = 0;
+  int max_simultaneous_tests;
+  struct ndtchild *tmp_ptr, *prev_ptr;
```

Comment:
Suggest `current_child`, `previous_child` (or `cur_child`, `prev_child`). I prefer to avoid `temp`/`tmp` unless it refers to a temporary storage location (e.g. we're going to do a swap). But outside of that, I feel like there are usually more meaningful names.